### ISSUE TITLE
Convenience function for wizards

### ISSUE THREAD

**Issue by jasongrout** [jupyter-widgets/ipywidgets#1725]
It would be interesting to explore making a simple "wizard" convenience function, that would take a sequence of functions. Each function would be executed inside of an Output widget, and we could put a button to advance to the next step automatically. I'm not sure what the inputs to the functions would be - perhaps we would just depend on global state or something.

**Comment by jasongrout** [comment]
This is probably an advanced sprinting issue - it's nice that it's self-contained, and fairly straightforward to get an initial prototype working.

**Comment by CharanyaR** [comment]
@jasongrout Can i take this up? ...If you can give me some pointers i can proceed on this

**Comment by pbugnion** [comment]
Definitely. @jasongrout may correct me, but I think the intention is to build something like this (if you'll excuse my lack of design skills):

<img width="1112" alt="screen shot 2017-11-01 at 09 03 04" src="https://user-images.githubusercontent.com/1392879/32267748-33d7f4d2-bee4-11e7-80b0-885eafca7931.png">

Basically, the user has a set of functions that they want to run. Each function mutates some kind of global state (in the example, functions mutate the internals of `MLPipeline`). The widget lets the user step through each function in the pipeline, keeping track of the output. In the initial implementation, we could just assume that each function takes no input parameters -- it just relies on some external state.

In terms of implementation, as Jason suggested, wrapping each function call in an output widget to capture stdout / stderr of that function seems sensible.

**Comment by CharanyaR** [comment]
@pbugnion thanks for the information, i will take a look at it and get back if any other information needed.

**Comment by jasongrout** [comment]
@pbugnion - that looks great. I was also thinking of an interface with tabs, one for each function. You have a class that maintains some internal state and has, say, methods A, B, and C. Calling a "run" method creates a tab widget labeled "A", puts a "Next" button and an output widget inside of it, then runs method A in the context of that output widget. Clicking the next button creates a new tab "B", a new output widget and next button, and runs method B in context of that output. A user can go back to a previous step by clicking the appropriate tab and restarting from there.

As a next step, perhaps each method can also specify some inputs that can be arranged using the interact layout logic, or perhaps each method can have an associated function that returns a widget full of controls for inputs.

**Comment by pbugnion** [comment]
I think an interface with tabs (rather than vertically stacked as in my drawing) makes a lot of sense, specially if some functions have a lot of output.

**Comment by CharanyaR** [comment]
@pbugnion @jasongrout I am working on it and i will update my progress early next week.

**Comment by BarbaraPr** [comment]
@CharanyaR thanks for working on the wizard widget. Is there a chance to use such an ipywidget in the near future?

**Comment by tslaton** [comment]
I've been working on something kind of similar. The whole thing below the cell pictured is a widget:
![](https://i.imgur.com/fNXepfT.png)

In the screenshot above, the intent was for this widget to be able to run an entire notebook (eg., by reading it as a file and directly parsing out the code) with the parameters and values in each tab as separate "experiments" that would affect the entire notebook as if they were global variables in the notebook (kind of like [Papermill](https://github.com/nteract/papermill)).

I'm also interested in a similar concept where a widget with a similar UI to the pictured one would be running a particular function within the notebook with the various sets of parameters as opposed to an entire notebook. Currently, I am assuming I can rely on the user to establish a 1:1 mapping between a cell and the function they want to explore.  Given this, I can convert the cell contents into a function and/or execute that function after parsing its source code from the notebook, as a file.

Does this issue already contain the latest thoughts from the Jupyter community on this idea of "wizards"? Is parsing the notebook from within the notebook a bad move? Any other advice?
